Adding ECON-T rawdata decoding corresponding to trigger primitive generation using STC mode

indra-ehep opened the pull request:
#### PR description:

  -  The file DQM/HGCal/plugins/HGCalTriggerClient.cc has been updated to include ECON-T decoding for STC mode
  -  It does not depend on other PR
  -  Please refer to the discussion in the "Trigger blocks" section in page 2 of https://indico.cern.ch/event/1341670/contributions/5648441/attachments/2742123/4771994/HGCAL_DPG_RawDataHandling_Feedback_2023Oct30.pdf

#### PR validation:

  - No separate verification is required, however performed "scram build code-checks ; scram build code-format".

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

  - no backporting is required

Commit:
Adding ECON-T rawdata decoding corresponding to trigger primitive generation using STC mode

adavidzh commented on DQM/HGCal/plugins/HGCalTriggerClient.cc:
```diff
@@ -45,8 +45,16 @@ class HGCalTriggerClient : public DQMEDAnalyzer {
+  MonitorElement *h_trigtime, *h_trigtype, *h_ECONTRawDataErrors;
+
+  // ------------ member function ------------
+  bool is0xFECAFE(const uint32_t word);
```
I think that all mentions to `0xFECAFE` in names should be replaced with `markerWord` or some meaningful such. The actual value of the constant is not relevant to name functions/methods.

indra-ehep replied:
Yes we can make those changes. What do you think about the word "blockSeparator" as a replacement for 0xfecafe ? As it separates different raw data blocks, e.g. unpacker input, output, scintillator....

indra-ehep replied:
Necessary modifications have been applied (see [8165d07](https://github.com/CMS-HGCAL/cmssw/pull/86/commits/8165d077af54e5ab2fcceb6c538b92669bbe113d))

adavidzh commented on DQM/HGCal/plugins/HGCalTriggerClient.cc:
```diff
@@ -57,24 +65,136 @@ class HGCalTriggerClient : public DQMEDAnalyzer {
+
+//Checks of matches with 0xFECAFE
+bool HGCalTriggerClient::is0xFECAFE(const uint32_t word) {
+  if (((word >> 8) & 0xFFFFFF) == 16698110)
```
So:
- By construction, `word>>8` is identical to `word>>8 & 0xFFFFFF`.
- `0xFECAFE` is not even used in the code?

I suggest `if( (word>>8) == 0xFECAFE )` or `if( word & 0xFFFFFF00 == 0xFECAFE00)`.

adavidzh replied:
Also, the `if` is not needed, e.g. `return ( word & 0xFFFFFF00 == 0xFECAFE00 )`.

indra-ehep replied:
Necessary modifications have been applied (see [8165d07](https://github.com/CMS-HGCAL/cmssw/pull/86/commits/8165d077af54e5ab2fcceb6c538b92669bbe113d))
However, kept the if structure.

adavidzh commented on DQM/HGCal/plugins/HGCalTriggerClient.cc:
```diff
@@ -84,103 +204,434 @@ void HGCalTriggerClient::analyze(const edm::Event& iEvent, const edm::EventSetup
+  bool isEngE0 = false;
+  bool isEngE1 = false;
+
+  if (trigType == 0x0001)
```
This should be a `switch` and `trigType` an instance of an `enum` so that you can benefit from [compile-time warnings and optimisations](https://brevzin.github.io/c++/2019/08/01/enums-default/).

indra-ehep replied:
Necessary modifications have been applied (see [8165d07](https://github.com/CMS-HGCAL/cmssw/pull/86/commits/8165d077af54e5ab2fcceb6c538b92669bbe113d))

adavidzh commented on DQM/HGCal/plugins/HGCalTriggerClient.cc:
```diff
@@ -84,103 +204,434 @@ void HGCalTriggerClient::analyze(const edm::Event& iEvent, const edm::EventSetup
+  uint32_t packet_locations[12];
+  uint64_t packet_energies[12];
+
+  for (int iect = 0; iect < 2; iect++) {
```
This is not needed if you [use `{0}`](https://stackoverflow.com/questions/31114644/initialize-multidimensional-array-with-zeros).

indra-ehep replied:
Necessary modifications have been applied (see [8165d07](https://github.com/CMS-HGCAL/cmssw/pull/86/commits/8165d077af54e5ab2fcceb6c538b92669bbe113d))

adavidzh commented on DQM/HGCal/plugins/HGCalTriggerClient.cc:
```diff
@@ -84,103 +204,434 @@ void HGCalTriggerClient::analyze(const edm::Event& iEvent, const edm::EventSetup
+  }
+
+  //check bx
+  for (int iect = 0; iect < 2; iect++) {
```
This is the first of a series of triple-nested `for` loops. Does it make sense to to the same loop multiple times or could all checks be done in one (triple-nested) loop?

indra-ehep replied:
Necessary modifications have been applied (see [8165d07](https://github.com/CMS-HGCAL/cmssw/pull/86/commits/8165d077af54e5ab2fcceb6c538b92669bbe113d))

adavidzh commented on DQM/HGCal/plugins/HGCalTriggerClient.cc:
```diff
@@ -84,103 +204,434 @@ void HGCalTriggerClient::analyze(const edm::Event& iEvent, const edm::EventSetup
+    }
+  }
+
+  if (isSTCNE0)
```
Same comment as above on `switch`.

indra-ehep replied:
These are different than the switch/case for L1a trigger type. The trigger type is unique for a given event. However, more than one errors of STC location, indices, bx or energy can be present for a given event. So if switch is used for one type others will not be checked. Therefore it is not applied at present. I press the "resolve" button below. Please reopen it if I have missed the point.

adavidzh replied:
I see your point; I was mislead by the code structure.

Imagine next time you want to add a new one bin to that histogram: you'd have to add something in 3 places.
In this sort of case, I'd have a `map<string, int>`: the keys are all the errors you declare and then you can increment and fill in an automated way.

Commit:
Updating ECON-T rawdata integration following the suggestions

indra-ehep commented:
@adavidzh   No need check the code that I have just pushed now, since I have not checked your later comments.

Commit:
The code has been updated with the first round of suggestions of Andre

indra-ehep commented:
@adavidzh Thanks for your review. I have addressed all of the above comments. Necessary modifications have been applied at [8165d07](https://github.com/CMS-HGCAL/cmssw/pull/86/commits/8165d077af54e5ab2fcceb6c538b92669bbe113d). I have also incorporated the suggestion of @pfs related to 2D error histogram.
Let me know, if there is any further comment or it is good to merge.

adavidzh commented on DQM/HGCal/plugins/HGCalTriggerClient.cc:
```diff
@@ -239,17 +240,25 @@ void HGCalTriggerClient::analyze(const edm::Event& iEvent, const edm::EventSetup
+  case Regular:
     h_trigtype->Fill(4);
-
+    break;
```
Consider adding a `default` statement. There are several possible [conventions](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-default).

indra-ehep replied:
The default option has been added for unknown trigger types (please see [def7b37](https://github.com/CMS-HGCAL/cmssw/pull/86/commits/def7b3763f165986b4342110f3f0dc38856f0abf) ).

adavidzh commented on DQM/HGCal/plugins/HGCalTriggerClient.cc:
```diff
@@ -66,6 +66,8 @@ class HGCalTriggerClient : public DQMEDAnalyzer {
   std::map<MonitorKey_t, MonitorKey_t> module_keys_;
   std::map<uint32_t, HGCalSiCellChannelInfo> eleidtoSiInfo_;
 
+  enum TriggerType { Phys=0x0001, Calib=0x0002, Random=0x0004, Soft=0x0008, Regular=0x0010};
```
[`enum class` is preferred](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Renum-class).

indra-ehep replied:
Now defined as class (please see [def7b37](https://github.com/CMS-HGCAL/cmssw/pull/86/commits/def7b3763f165986b4342110f3f0dc38856f0abf) ).

Commit:
The code has been updated with the second round of suggestions of Andre

indra-ehep commented:
@adavidzh : Thanks for second round of comments. I have included them in latest update (please see [def7b37](https://github.com/CMS-HGCAL/cmssw/pull/86/commits/def7b3763f165986b4342110f3f0dc38856f0abf) ).

adavidzh commented:
> @adavidzh : Thanks for second round of comments. I have included them in latest update (please see [def7b37](https://github.com/CMS-HGCAL/cmssw/pull/86/commits/def7b3763f165986b4342110f3f0dc38856f0abf) ).

Thanks. I am not an approver here, just an interested party.

indra-ehep commented:
@adavidzh Do you know who will merge the changes ?

pfs commented:
@adavidzh do you want to do merge - after all you did the most careful check of the code :)

@indra-ehep Next step is to create the "ECONT Raw to Digi producer" to fill the data format to persist. On this front is it useful to contact Emyr, Jean-Baptiste ?

adavidzh commented:
@pfs, I did not review the code. I just suggested a couple of improvements :P

indra-ehep commented:
@pfs @adavidzh : Can the official review be started ? So that the merging is possible.
@pfs : I am in touch with Emyr and will be in contact with Jean-Baptiste.

pfs commented:
Hi @indra-ehep I'm ok with merging.
